jersey/src/main/java/org/pac4j/jax/rs/jersey/features/Pac4JValueFactoryProvider.java

```diff
@@ -165,69 +159,94 @@ public Binder() {
             bind(profile).to(ProfileFactoryBuilder.class);
             bind(optProfile).to(OptionalProfileFactoryBuilder.class);
-            bind(manager).to(ProfileManagerFactoryBuilder.class);
 
-            bind(Pac4JProfileValueFactoryProvider.class).to(ValueFactoryProvider.class).in(Singleton.class);
+            if(manager == null){
+                bind(DefaultProfileManagerFactoryBuilder.class)
+                    .to(ProfileManagerFactoryBuilder.class)
+                ;
+            } else {
+                bind(manager).to(ProfileManagerFactoryBuilder.class);
+            }
+
+            bind(Pac4JProfileValueFactoryProvider.class).to(ValueParamProvider.class).in(Singleton.class);
 
-            bind(ProfileInjectionResolver.class).to(new TypeLiteral<InjectionResolver<Pac4JProfile>>() {
-            }).in(Singleton.class);
-            bind(ProfileManagerInjectionResolver.class).to(new TypeLiteral<InjectionResolver<Pac4JProfileManager>>() {
-            }).in(Singleton.class);
+            bind(ProfileInjectionResolver.class)
+                .to(new GenericType<InjectionResolver<Pac4JProfile>>(){})
+                .in(Singleton.class);
+
+            bind(ProfileManagerInjectionResolver.class)
+                .to(new GenericType<InjectionResolver<Pac4JProfileManager>>(){})
+                .in(Singleton.class);
         }
     }
 
-    static class ProfileManagerValueFactory extends AbstractContainerRequestValueFactory<ProfileManager<CommonProfile>>
-            implements ProfileManagerFactory {
-
+    static class ProfileManagerValueFactory implements ProfileManagerFactory{
         @Context
-        Providers providers;
+        private Providers providers;
+
+        ProfileManagerValueFactory(Providers providers) {
+            this.providers = providers;
+        }
+
+        @Override
+        public ProfileManager<CommonProfile> apply(ContainerRequest containerRequest) {
+            return new RequestProfileManager(new RequestJaxRsContext(providers, containerRequest))
+                .profileManager();
+        }
+    }
 
+    static class ProfileValueFactory implements ProfileFactory {
         @Override
-        public ProfileManager<CommonProfile> provide() {
-            return new RequestProfileManager(new RequestJaxRsContext(providers, getContainerRequest()))
-                    .profileManager();
+        public CommonProfile apply(ContainerRequest containerRequest) {
+            return optionalProfile(containerRequest)
+                .orElseThrow(() -> {
+                    LOG.debug("Cannot inject a Pac4j profile into an unauthenticated request, responding with 401");
+                    return new WebApplicationException(401);
+                });
         }
     }
 
-    static class ProfileValueFactory extends AbstractContainerRequestValueFactory<CommonProfile>
-            implements ProfileFactory {
+    static class OptionalProfileValueFactory implements OptionalProfileFactory {
```

Comment:
any reason why you don't rely on injection here to get the request instead of making it explicit in the interface?
I liked the idea of having an interface agnostic to the where the request was coming from… but maybe it's not possible anymore?

Reply:
Well, the main reason is that I could not find a way to make it work. A secondary reason is that the origin of the request as a parameter comes directly from the AbstractValueParamProvider.
To me it felt like the way to go, but I'm no expert.